File: answers/answers.java
import java.util.*;

public class CommonCodingProblems {
    
    // 1. Reverse a String
    public static String reverseString(String str) {
        // Use StringBuilder to reverse the string efficiently
        return new StringBuilder(str).reverse().toString();
    }

    // 2. Check for Palindrome (String remains same when reversed)
    public static boolean isPalindrome(String str) {
        String reversed = reverseString(str);
        return str.equals(reversed); // Compare original and reversed strings
    }

    // 3. Find the Factorial of a number using Recursion
    public static int factorial(int n) {
        if (n == 0 || n == 1) return 1; // Base case
        return n * factorial(n - 1); // Recursive case
    }

    // 4. Fibonacci Sequence: Print first n Fibonacci numbers
    public static void fibonacci(int n) {
        int a = 0, b = 1;
        System.out.print(a + " " + b); // Print first two numbers
        for (int i = 2; i < n; i++) {
            int next = a + b;
            System.out.print(" " + next);
            a = b;
            b = next;
        }
        System.out.println();
    }

    // 5. Check if a number is Prime
    public static boolean isPrime(int n) {
        if (n < 2) return false; // Prime numbers are >= 2
        for (int i = 2; i <= Math.sqrt(n); i++) { // Check divisibility up to sqrt(n)
            if (n % i == 0) return false; // Not prime if divisible
        }
        return true;
    }

    // 6. Find the Largest and Smallest Element in an Array
    public static int[] findMinMax(int[] arr) {
        int min = Integer.MAX_VALUE, max = Integer.MIN_VALUE;
        for (int num : arr) {
            if (num < min) min = num;
            if (num > max) max = num;
        }
        return new int[]{min, max}; // Return both min and max values
    }

    // 7. Find the Second Largest Number in an Array
    public static int secondLargest(int[] arr) {
        int largest = Integer.MIN_VALUE, second = Integer.MIN_VALUE;
        for (int num : arr) {
            if (num > largest) { 
                second = largest; 
                largest = num; 
            } else if (num > second && num != largest) { 
                second = num; 
            }
        }
        return second;
    }

    // 8. Remove Duplicates from an Array using Set
    public static int[] removeDuplicates(int[] arr) {
        Set<Integer> set = new LinkedHashSet<>();
        for (int num : arr) set.add(num);
        return set.stream().mapToInt(i -> i).toArray(); // Convert back to array
    }

    // 9. Check if Two Strings Are Anagrams (same letters, different order)
    public static boolean areAnagrams(String str1, String str2) {
        char[] arr1 = str1.toCharArray();
        char[] arr2 = str2.toCharArray();
        Arrays.sort(arr1);
        Arrays.sort(arr2);
        return Arrays.equals(arr1, arr2);
    }

    // 10. Find the First Non-Repeating Character in a String
    public static char firstNonRepeatingChar(String str) {
        Map<Character, Integer> countMap = new LinkedHashMap<>();
        for (char c : str.toCharArray()) {
            countMap.put(c, countMap.getOrDefault(c, 0) + 1); // Count occurrences
        }
        for (char c : str.toCharArray()) {
            if (countMap.get(c) == 1) return c; // First unique character
        }
        return '\0'; // Return null character if none found
    }

    // 11. Bubble Sort Algorithm
    public static void bubbleSort(int[] arr) {
        int n = arr.length;
        for (int i = 0; i < n - 1; i++) { // Iterate over array
            for (int j = 0; j < n - i - 1; j++) { // Compare adjacent elements
                if (arr[j] > arr[j + 1]) { // Swap if needed
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                }
            }
        }
    }

    // 12. Binary Search (for sorted array)
    public static int binarySearch(int[] arr, int target) {
        int left = 0, right = arr.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (arr[mid] == target) return mid; // Found the element
            if (arr[mid] < target) left = mid + 1;
            else right = mid - 1;
        }
        return -1; // Element not found
    }

    // 13. Find the Missing Number in an Array (1 to n)
    public static int findMissingNumber(int[] arr, int n) {
        int sum = n * (n + 1) / 2; // Expected sum formula
        for (int num : arr) sum -= num; // Subtract elements from sum
        return sum; // Missing number
    }

    // 14. Find the Intersection of Two Arrays
    public static int[] arrayIntersection(int[] arr1, int[] arr2) {
        Set<Integer> set1 = new HashSet<>();
        Set<Integer> result = new HashSet<>();
        for (int num : arr1) set1.add(num);
        for (int num : arr2) if (set1.contains(num)) result.add(num);
        return result.stream().mapToInt(i -> i).toArray();
    }

    // 15. Reverse a Linked List
    static class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
    }
    public static ListNode reverseLinkedList(ListNode head) {
        ListNode prev = null;
        while (head != null) {
            ListNode next = head.next;
            head.next = prev; // Reverse link
            prev = head;
            head = next;
        }
        return prev; // New head
    }

    // 16. Detect a Cycle in a Linked List (Floyd’s Cycle Detection)
    public static boolean hasCycle(ListNode head) {
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) return true; // Cycle detected
        }
        return false;
    }

    // 17. Implement a Stack Using Arrays
    static class Stack {
        private List<Integer> list = new ArrayList<>();
        public void push(int val) { list.add(val); }
        public int pop() { return list.isEmpty() ? -1 : list.remove(list.size() - 1); }
        public int peek() { return list.isEmpty() ? -1 : list.get(list.size() - 1); }
    }

    // 18. Implement a Queue Using Two Stacks
    static class QueueUsingStack {
        private Stack<Integer> stack1 = new Stack<>();
        private Stack<Integer> stack2 = new Stack<>();
        public void enqueue(int val) { stack1.push(val); }
        public int dequeue() {
            if (stack2.isEmpty()) while (!stack1.isEmpty()) stack2.push(stack1.pop());
            return stack2.isEmpty() ? -1 : stack2.pop();
        }
    }

    // 19. Find the Longest Substring Without Repeating Characters
    public static int longestUniqueSubstring(String s) {
        int maxLength = 0, left = 0;
        Map<Character, Integer> seen = new HashMap<>();
        for (int right = 0; right < s.length(); right++) {
            if (seen.containsKey(s.charAt(right))) left = Math.max(left, seen.get(s.charAt(right)) + 1);
            seen.put(s.charAt(right), right);
            maxLength = Math.max(maxLength, right - left + 1);
        }
        return maxLength;
    }

    // 20. Solve the Two Sum Problem
    public static int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(target - nums[i])) return new int[]{map.get(target - nums[i]), i};
            map.put(nums[i], i);
        }
        return new int[]{};
    }

    public static void main(String[] args) {
        System.out.println("Java Implementation of 20 Common Coding Problems");
    }
}

File: answers/answers.py
"""
Common Algorithm Questions in Python
"""

# 1. Reverse a String
# This function takes a string and returns it reversed using slicing.
def reverse_string(s):
    return s[::-1]

# 2. Check for Palindrome
# A palindrome is a string that reads the same forward and backward.
# This function checks if the given string is a palindrome.
def is_palindrome(s):
    return s == s[::-1]

# 3. Find the Factorial (Recursive)
# The factorial of a number is the product of all positive integers up to that number.
# This function calculates factorial recursively.
def factorial(n):
    if n == 0 or n == 1:
        return 1  # Base case
    return n * factorial(n - 1)  # Recursive case

# 4. Fibonacci Sequence
# This function generates the first `n` numbers of the Fibonacci sequence.
def fibonacci(n):
    fib = [0, 1]  # First two numbers
    for i in range(2, n):  # Generate the sequence
        fib.append(fib[i - 1] + fib[i - 2])
    return fib[:n]

# 5. Check for Prime Number
# A prime number is a number greater than 1 that has no divisors other than 1 and itself.
def is_prime(n):
    if n < 2:
        return False  # Numbers less than 2 are not prime
    for i in range(2, int(n ** 0.5) + 1):  # Check divisibility up to sqrt(n)
        if n % i == 0:
            return False
    return True

# 6. Find the Largest and Smallest Element in an Array
# This function returns both the smallest and largest number in the given array.
def find_min_max(arr):
    return min(arr), max(arr)

# 7. Find the Second Largest Number in an Array
# This function finds the second largest number by sorting unique elements.
def second_largest(arr):
    unique_arr = list(set(arr))  # Remove duplicates
    unique_arr.sort(reverse=True)  # Sort in descending order
    return unique_arr[1] if len(unique_arr) > 1 else None  # Return second largest if available

# 8. Remove Duplicates from an Array
# This function removes duplicate values by converting the list to a set.
def remove_duplicates(arr):
    return list(set(arr))

# 9. Check if Two Strings Are Anagrams
# An anagram is when two words have the same characters but in a different order.
from collections import Counter
def are_anagrams(s1, s2):
    return Counter(s1) == Counter(s2)  # Compare character counts

# 10. Find the First Non-Repeating Character in a String
# This function finds the first character that appears only once.
def first_non_repeating(s):
    char_count = Counter(s)  # Count occurrences of each character
    for char in s:
        if char_count[char] == 1:  # Return first unique character
            return char
    return None  # Return None if all characters repeat

# 11. Implement Bubble Sort
# Bubble Sort is a simple sorting algorithm that repeatedly swaps adjacent elements if they are in the wrong order.
def bubble_sort(arr):
    n = len(arr)
    for i in range(n):
        for j in range(0, n-i-1):  # Last i elements are already sorted
            if arr[j] > arr[j+1]:  # Swap if needed
                arr[j], arr[j+1] = arr[j+1], arr[j]
    return arr

# 12. Binary Search Implementation
# Binary search efficiently finds an element in a sorted array by repeatedly dividing it in half.
def binary_search(arr, target):
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid  # Element found
        elif arr[mid] < target:
            left = mid + 1  # Search right half
        else:
            right = mid - 1  # Search left half
    return -1  # Element not found

# 13. Find the Missing Number in an Array of 1 to n
# This function finds a missing number in a sequence by using the sum formula.
def missing_number(arr, n):
    return (n * (n + 1) // 2) - sum(arr)  # Expected sum - actual sum

# 14. Find the Intersection of Two Arrays
# This function finds the common elements between two lists.
def intersection(arr1, arr2):
    return list(set(arr1) & set(arr2))  # Use set intersection

# 15. Reverse a Linked List
# This function reverses a singly linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverse_linked_list(head):
    prev, curr = None, head
    while curr:
        next_node = curr.next  # Store next node
        curr.next = prev  # Reverse the link
        prev, curr = curr, next_node  # Move forward
    return prev  # New head of reversed list

# 16. Detect a Cycle in a Linked List
# This function uses Floyd's cycle detection algorithm to detect a loop.
def has_cycle(head):
    slow, fast = head, head
    while fast and fast.next:
        slow, fast = slow.next, fast.next.next  # Move slow by 1 and fast by 2
        if slow == fast:  # If they meet, there is a cycle
            return True
    return False  # No cycle detected

# 17. Implement a Stack Using Arrays
# A stack follows Last-In-First-Out (LIFO) principle.
class Stack:
    def __init__(self):
        self.stack = []

    def push(self, val):
        self.stack.append(val)  # Add element to the stack

    def pop(self):
        return self.stack.pop() if self.stack else None  # Remove last element

    def peek(self):
        return self.stack[-1] if self.stack else None  # Return last element

    def is_empty(self):
        return len(self.stack) == 0  # Check if stack is empty

# 18. Implement a Queue Using Stacks
# A queue follows First-In-First-Out (FIFO) principle.
class QueueUsingStacks:
    def __init__(self):
        self.stack1 = []  # Primary stack
        self.stack2 = []  # Secondary stack for reversing order

    def enqueue(self, val):
        self.stack1.append(val)  # Push new element onto stack1

    def dequeue(self):
        if not self.stack2:  # If stack2 is empty, move elements from stack1
            while self.stack1:
                self.stack2.append(self.stack1.pop())
        return self.stack2.pop() if self.stack2 else None  # Pop from stack2

# 19. Find the Longest Substring Without Repeating Characters
# This function finds the length of the longest substring without repeating characters.
def longest_unique_substring(s):
    char_index = {}
    left, max_len = 0, 0
    for right in range(len(s)):
        if s[right] in char_index:
            left = max(left, char_index[s[right]] + 1)  # Move left pointer
        char_index[s[right]] = right  # Store index of current character
        max_len = max(max_len, right - left + 1)  # Update max length
    return max_len

# 20. Solve the "Two Sum" Problem
# Given an array and a target sum, this function finds two numbers that add up to the target.
def two_sum(nums, target):
    num_map = {}
    for i, num in enumerate(nums):
        complement = target - num  # Find the difference
        if complement in num_map:
            return [num_map[complement], i]  # Return the indices
        num_map[num] = i  # Store current number's index
    return []